[iOS] The selection highlight boxes mess up when the content switch between inflow and out-of-flow

cathiechen opened the pull request:
#### 834e9831da4db3819ec6547aaba1f98b5bd89a32
<pre>
[iOS] The selection highlight boxes mess up when the content switch between inflow and out-of-flow
<a href="https://bugs.webkit.org/show_bug.cgi?id=269723">https://bugs.webkit.org/show_bug.cgi?id=269723</a>

Reviewed by Wenson Hsieh.

In iOS, the selection highlight boxes are coalesced. But if the selected content switch between inflow and
out-of-flow, for instance, the content contains text with positioned ancestor and text from inflow,
the combined box could include unexpected area.
To fix this issue, we stop coalescing the highlight boxes of out-of-flow positioned content to others.

* LayoutTests/editing/selection/ios/select-content-from-different-flow-001-expected.txt: Added.
* LayoutTests/editing/selection/ios/select-content-from-different-flow-001.html: Added.
* LayoutTests/editing/selection/ios/select-content-from-different-flow-002-expected.txt: Added.
* LayoutTests/editing/selection/ios/select-content-from-different-flow-002.html: Added.
* LayoutTests/editing/selection/ios/select-content-from-different-flow-003-expected.txt: Added.
* LayoutTests/editing/selection/ios/select-content-from-different-flow-003.html: Added.
* LayoutTests/editing/selection/ios/select-content-from-different-flow-004-expected.txt: Added.
* LayoutTests/editing/selection/ios/select-content-from-different-flow-004.html: Added.
* LayoutTests/editing/selection/ios/select-content-from-different-flow-005-expected.txt: Added.
* LayoutTests/editing/selection/ios/select-content-from-different-flow-005.html: Added.
* Source/WebCore/platform/ios/SelectionGeometry.h:
(WebCore::SelectionGeometry::separateFromPreviousLine const): Return true if geometry requires to separate lines.
(WebCore::SelectionGeometry::setSeparateFromPreviousLine):
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::currentNodeRequiresToSeparateLines): Return true if current node is out-of-flow positioned.
(WebCore::hasAncestorRequiresToSeparateLines): Check if the node has an out-of-flow ancestor within the stayWithin subtree.
(WebCore::previousNodeRequiresToSeparateLines):
(WebCore::RenderObject::collectSelectionGeometriesInternal): Distinguish if current renderer and previous renderer
have switched between inflow and out-of-flow. If current renderer is positioned, or if the previous renderer
has a positioned ancestor, then they are in different flows. Do not coalesce lines that are not in same flow.

Canonical link: <a href="https://commits.webkit.org/291696@main">https://commits.webkit.org/291696@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/24b4c68566b6877ded5f01759ca5b486d9de1b55

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/93725 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/13299 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/3036 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/98731 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/44251 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/95775 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/13594 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/21741 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/71551 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/28922 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/96727 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/10119 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/84709 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/51885 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/9800 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/2336 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/43566 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/80073 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/2404 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/100765 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/20777 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/15159 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/80567 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/21029 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/80651 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/79904 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19876 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/24453 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/1807 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/13933 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/20761 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/25939 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/20448 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/23908 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/22189 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on Source/WebCore/rendering/RenderObject.cpp:
```diff
@@ -2642,6 +2642,17 @@ static void adjustTextDirectionForCoalescedGeometries(const SimpleRange& range,
     }
 }
 
+static bool isDescendantOutOfFlow(RenderObject* descendant, const RenderObject* ancestor)
```
isDescendant`Of`OutOfFlow?

alanbaradlay commented on Source/WebCore/rendering/RenderObject.cpp:
```diff
@@ -2642,6 +2642,17 @@ static void adjustTextDirectionForCoalescedGeometries(const SimpleRange& range,
 
+static bool isDescendantOutOfFlow(RenderObject* descendant, const RenderObject* ancestor)
+{
+    for (; descendant; descendant = descendant->containingBlock()) {
```
This is slightly confusing as you walk **_up_** on the **_ancestor_** chain and yet the variable is called **_descendant_**.

alanbaradlay replied:
Is this a containing **_block_** or a container based helper? i.e. is it okay to skip over inflow positioned inline boxes?

cathiechen replied:
Yes, true, skipping inline boxes is wrong, I will use `parent()` here, and add a test for it.

alanbaradlay commented on Source/WebCore/rendering/RenderObject.cpp:
```diff
@@ -2642,6 +2642,17 @@ static void adjustTextDirectionForCoalescedGeometries(const SimpleRange& range,
+    for (; descendant; descendant = descendant->containingBlock()) {
+        if (descendant->isOutOfFlowPositioned())
+            return true;
+        if (descendant == ancestor)
```
so ancestor acts as stayWithin? Maybe name it accordingly?

alanbaradlay commented on Source/WebCore/rendering/RenderObject.cpp:
```diff
@@ -2650,10 +2661,21 @@ auto RenderObject::collectSelectionGeometriesInternal(const SimpleRange& range)
     bool containsDifferentWritingModes = false;
     bool hasLeftToRightText = false;
     bool hasRightToLeftText = false;
+    bool isInDifferentFlow = false;
```
what does **_flow_** mean in this context? Different RenderBlockFlow? Different formatting context subtree? Inflow vs. out-of-flow? something else?

cathiechen replied:
It means inflow vs. out-of-flow, how about `hasFlowTransition`?

cathiechen replied:
We replaced it with `needsToSeparateBoundary` here, and reveal the position info in functions `currentNodeRequiresToSeparateBoundary` and `lastNodeRequiresToSeparateBoundary`. Hope it looks clearer now:)

whsieh commented on Source/WebCore/rendering/RenderObject.cpp:
```diff
@@ -2650,10 +2673,21 @@ auto RenderObject::collectSelectionGeometriesInternal(const SimpleRange& range)
     bool hasLeftToRightText = false;
     bool hasRightToLeftText = false;
+    bool needsToSeparateBoundary = false;
+    RenderObject* lastRenderer = nullptr;
```
You should probably use `WeakPtr` here instead of a raw pointer.

(`CheckedPtr` may not be viable here because there's a potential call to `VisiblePosition::VisiblePosition` below, which may update layout and therefore destroy the renderer that `lastRenderer` points to).

cathiechen replied:
Done, thanks:)

whsieh commented on Source/WebCore/platform/ios/SelectionGeometry.h:
```diff
@@ -108,6 +110,7 @@ class SelectionGeometry {
     bool m_containsEnd { false };
     bool m_isHorizontal { true };
     bool m_isInFixedPosition { false };
+    bool m_separateBoundary { false };
```
I think this might be clearer as `bool m_isOutOfFlow` or `bool m_isOutOfFlowPositioned`.

cathiechen replied:
It seems we could reuse this flag for other scenarios, like multi-column.
OK, let's use `m_isOutOfFlowPositioned` now, then we may discuss again at the multi-column patch, which I will upload it to review soon.

cathiechen replied:
Hi @whsieh, `m_isOutOfFlowPositioned` seems not very accurate here. Because this flag is also true when current node is inflow and previous node is out-of-flow. This flag indicates something like position state change, static to absolute, or absolute to static.
How about `m_hasPositionStateChanged`?

whsieh replied:
I see, it seems I misunderstood that part. I suppose we could describe the intended behavior that results from setting the flag — something like `m_separateFromPreviousLine`

whsieh commented on Source/WebCore/rendering/RenderObject.cpp:
```diff
@@ -2645,6 +2645,29 @@ static void adjustTextDirectionForCoalescedGeometries(const SimpleRange& range,
     }
 }
 
+static bool currentNodeRequiresToSeparateBoundary(const RenderObject* currentRenderer)
```
…and then this would become `currentNodeIsOutOfFlow` or `currentNodeIsOutOfFlowPositioned`

whsieh commented on Source/WebCore/rendering/RenderObject.cpp:
```diff
@@ -2645,6 +2645,29 @@ static void adjustTextDirectionForCoalescedGeometries(const SimpleRange& range,
+
+static bool isDescendantOfOutOfFlow(RenderObject* descendant, const RenderObject* stayWithin)
+{
+    for (RenderObject* current = descendant; current; current = current->parent()) {
```
Nit - you could deploy smart pointers here too:

```suggestion
    for (CheckedPtr current = descendant; current; current = current->parent()) {
```

whsieh commented on Source/WebCore/rendering/RenderObject.cpp:
```diff
@@ -2645,6 +2645,29 @@ static void adjustTextDirectionForCoalescedGeometries(const SimpleRange& range,
+    return currentRenderer && currentRenderer->isOutOfFlowPositioned();
+}
+
+static bool isDescendantOfOutOfFlow(RenderObject* descendant, const RenderObject* stayWithin)
```
The name seems slightly ambiguous, since it could be interpreted as checking whether the given renderer has a descendant that is out of flow — maybe `hasOutOfFlowAncestor()` instead?

whsieh commented on Source/WebCore/rendering/RenderObject.cpp:
```diff
@@ -2728,7 +2766,12 @@ auto RenderObject::collectSelectionGeometriesInternal(const SimpleRange& range)
+                lastLineTop = std::numeric_limits<int>::max();
+                lastLineBottom = std::numeric_limits<int>::min();
+                lineTop = currentRectTop;
+                lineBottom = currentRectBottom;
```
Would we still want to coalesce selection geometries in the case where there's two adjacent selected text nodes, that are _both_ within the same out-of-flow positioned renderer?

cathiechen replied:
Yes, we will coalesce the geometries if both are in the same positioned renderer.

whsieh commented on LayoutTests/editing/selection/ios/select-content-from-different-flow-002.html:
```diff
@@ -0,0 +1,54 @@
+    shouldBe("selectionRects.length", "2");
+    staticLineTop = selectionRects[0].top;
+    absoluteLineBottom = selectionRects[1].top + selectionRects[0].height;
+    shouldBeTrue("staticLineTop > absoluteLineBottom");
```
```suggestion
    shouldBeGreaterThan("staticLineTop", "absoluteLineBottom");
```

whsieh commented on LayoutTests/editing/selection/ios/select-content-from-different-flow-001.html:
```diff
@@ -0,0 +1,53 @@
+    shouldBe("selectionRects.length", "2");
+    staticLineBottom = selectionRects[0].top + selectionRects[0].height;
+    absoluteLineTop = selectionRects[1].top;
+    shouldBeTrue("absoluteLineTop > staticLineBottom");
```
```suggestion
    shouldBeGreaterThan("absoluteLineTop", "staticLineBottom");
```

whsieh commented on LayoutTests/editing/selection/ios/select-content-from-different-flow-003.html:
```diff
@@ -0,0 +1,54 @@
+    shouldBe("selectionRects.length", "2");
+    absoluteLineBottom = selectionRects[0].top + selectionRects[0].height;
+    staticLineTop = selectionRects[1].top;
+    shouldBeTrue("absoluteLineBottom < staticLineTop");
```
```suggestion
    shouldBeGreater("staticLineTop", "absoluteLineBottom");
```

whsieh commented on LayoutTests/editing/selection/ios/select-content-from-different-flow-004.html:
```diff
@@ -0,0 +1,54 @@
+    shouldBe("selectionRects.length", "2");
+    absoluteLineBottom = selectionRects[0].top + selectionRects[0].height;
+    staticLineTop = selectionRects[1].top;
+    shouldBeTrue("absoluteLineBottom < staticLineTop");
```
```suggestion
    shouldBeGreater("staticLineTop", "absoluteLineBottom");
```

whsieh commented on LayoutTests/editing/selection/ios/select-content-from-different-flow-005.html:
```diff
@@ -0,0 +1,56 @@
+    shouldBe("selectionRects.length", "2");
+    firstLineBottom = selectionRects[0].top + selectionRects[0].height;
+    secondLineTop = selectionRects[1].top;
+    shouldBeTrue("firstLineBottom == secondLineTop");
```
```suggestion
    shouldBe("firstLineBottom", "secondLineTop");
```

whsieh commented on Source/WebCore/rendering/RenderObject.cpp:
```diff
@@ -2653,10 +2676,21 @@ auto RenderObject::collectSelectionGeometriesInternal(const SimpleRange& range)
+        if (currentNodeRequiresToSeparateBoundary(renderer.get()))
+            needsToSeparateBoundary = true;
+        else if (!needsToSeparateBoundary)
+            needsToSeparateBoundary = lastNodeRequiresToSeparateBoundary(previousRenderer.get(), renderer->previousSibling());
```
Might be more succinct this way:

```suggestion
        needsToSeparateBoundary = currentNodeRequiresToSeparateBoundary(renderer.get()) || lastNodeRequiresToSeparateBoundary(previousRenderer.get(), renderer->previousSibling());
```

whsieh reviewed (approved):
r=me with prior comments

cathiechen committed:
[iOS] The selection highlight boxes mess up when the content switch between inflow and out-of-flow
https://bugs.webkit.org/show_bug.cgi?id=269723

Reviewed by Wenson Hsieh.

In iOS, the selection highlight boxes are coalesced. But if the selected content switch between inflow and
out-of-flow, for instance, the content contains text with positioned ancestor and text from inflow,
the combined box could include unexpected area.
To fix this issue, we stop coalescing the highlight boxes of out-of-flow positioned content to others.

* LayoutTests/editing/selection/ios/select-content-from-different-flow-001-expected.txt: Added.
* LayoutTests/editing/selection/ios/select-content-from-different-flow-001.html: Added.
* LayoutTests/editing/selection/ios/select-content-from-different-flow-002-expected.txt: Added.
* LayoutTests/editing/selection/ios/select-content-from-different-flow-002.html: Added.
* LayoutTests/editing/selection/ios/select-content-from-different-flow-003-expected.txt: Added.
* LayoutTests/editing/selection/ios/select-content-from-different-flow-003.html: Added.
* LayoutTests/editing/selection/ios/select-content-from-different-flow-004-expected.txt: Added.
* LayoutTests/editing/selection/ios/select-content-from-different-flow-004.html: Added.
* LayoutTests/editing/selection/ios/select-content-from-different-flow-005-expected.txt: Added.
* LayoutTests/editing/selection/ios/select-content-from-different-flow-005.html: Added.
* Source/WebCore/platform/ios/SelectionGeometry.h:
(WebCore::SelectionGeometry::separateFromPreviousLine const): Return true if geometry requires to separate lines.
(WebCore::SelectionGeometry::setSeparateFromPreviousLine):
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::currentNodeRequiresToSeparateLines): Return true if current node is out-of-flow positioned.
(WebCore::hasAncestorRequiresToSeparateLines): Check if the node has an out-of-flow ancestor within the stayWithin subtree.
(WebCore::previousNodeRequiresToSeparateLines):
(WebCore::RenderObject::collectSelectionGeometriesInternal): Distinguish if current renderer and previous renderer
have switched between inflow and out-of-flow. If current renderer is positioned, or if the previous renderer
has a positioned ancestor, then they are in different flows. Do not coalesce lines that are not in same flow.

Canonical link: https://commits.webkit.org/291696@main
